### PR TITLE
[MPS] fix compiling of SDPA producing nan results

### LINKED ISSUE #171764: torch.compile results in loss nan while it converges without torch.compile on mac cpu and mps
### 🐛 Describe the bug

torch.compile(model) on my macbook m1 pro, results in loss naN where as without torch.compile(model) the loss converges as expected.

``` Python
import os
import math
import time
import inspect
from dataclasses import dataclass
import torch
import torch.nn as nn
from torch.nn import functional as F

# --------------------------------------------------------------------------------------------------------

class CausalSelfAttention(nn.Module):

  def __init__(self, config):
    super().__init__()
    assert config.n_embd % config.n_head == 0
    
    self.c_attn = nn.Linear(config.n_embd, 3 * config.n_embd)
    self.c_proj = nn.Linear(config.n_embd, config.n_embd)
    self.c_proj.NANOGPT_SCALE_INIT = 1

    self.n_head = config.n_head
    self.n_embd = config.n_embd

    self.register_buffer('bias', torch.tril(torch.ones(config.block_size, config.block_size)).view(1, 1, config.block_size, config.block_size))

  def forward(self, x):
    B, T, C = x.size()
    qkv = self.c_attn(x)
    q, k, v = qkv.split(self.n_embd, dim=2)

    q = q.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)
    k = k.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)
    v = v.view(B, T, self.n_head, C // self.n_head).transpose(1, 2)

    y = F.scaled_dot_product_attention(q, k, v, is_causal=True) 
    y = y.transpose(1, 2).contiguous().view(B, T, C) 
    y = self.c_proj(y)
    return y


class MLP(nn.Module):

  def __init__(self, config):
    super().__init__()
    self.c_fc = nn.Linear(config.n_embd,  4 * config.n_embd)
    self.gelu = nn.GELU(approximate='tanh')
    self.c_proj = nn.Linear(4 * config.n_embd, config.n_embd)
    self.c_proj.NANOGPT_SCALE_INIT = 1

  def forward(self, x):
    x = self.c_fc(x)
    x = self.gelu(x)
    x = self.c_proj(x)
    return x


class Block(nn.Module):

  def __init__(self, config):
    super().__init__()
    self.ln_1 = nn.LayerNorm(config.n_embd)
    self.attn = CausalSelfAttention(config)
    self.ln_2 = nn.LayerNorm(config.n_embd)
    self.mlp = MLP(config)

  
  def forward(self, x):
    x = x + self.attn(self.ln_1(x))
    x = x + self.mlp(self.ln_2(x)) 
    return x

@dataclass
class GPTConfig:
  block_size: int = 1024        
  vocab_size: int = 50257       
  n_layer: int = 12             
  n_head: int = 12              
  n_embd: int = 768             

class GPT(nn.Module):

  def __init__(self, config):
    super().__init__()
    self.config = config

    self.transformer = nn.ModuleDict(dict(
      wte = nn.Embedding(config.vocab_size, config.n_embd), 
      wpe = nn.Embedding(config.block_size, config.n_embd), 
      h = nn.ModuleList([Block(config) for _ in range(config.n_layer)]), 
      ln_f = nn.LayerNorm(config.n_embd) 
    ))
    self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
    self.transformer.wte.weight = self.lm_head.weight
    self.apply(self._init_weights)

  
  def _init_weights(self, module):
    
    if isinstance(module, nn.Linear):
      std = 0.02

      if hasattr(module, 'NANOGPT_SCALE_INIT'):
        std *= (2 * self.config.n_layer) ** -0.5 

      torch.nn.init.normal_(module.weight, mean=0.0, std=std)
      if module.bias is not None:
        torch.nn.init.zeros_(module.bias)
      
    elif isinstance(module, nn.Embedding):
      torch.nn.init.normal_(module.weight, mean=0.0, std=0.02)


  def forward(self, idx, targets=None):
  
    B, T = idx.size()
    assert T <= self.config.block_size, f'Cannot forward sequence of length {T}, block size is only {self.config.block_size}'

    pos = torch.arange(0, T, dtype=torch.long, device=idx.device)
    pos_emb = self.transformer.wpe(pos) 
    tok_emb = self.transformer.wte(idx) 
    x = tok_emb + pos_emb 
    
    for block in self.transformer.h:
       x = block(x)

    
    x = self.transformer.ln_f(x)
    logits = self.lm_head(x)
    loss = None

    if targets is not None:
      loss = F.cross_entropy(logits.view(-1, logits.size(-1)), targets.view(-1))
  
    return logits, loss

# --------------------------------------------------------------------------------------------------------
import tiktoken

class DataLoaderLite:
   
  def __init__(self, B, T):
    self.B = B
    self.T = T

    with open('./data/input.txt', 'r') as f:
      text = f.read()
  
    enc = tiktoken.get_encoding('gpt2')
    tokens = enc.encode(text, allowed_special={'<|endoftext|>'})
    self.tokens = torch.tensor(tokens)
    print(f'loaded {len(self.tokens)} tokens')
    print(f'1 epoch = {len(self.tokens) // (B * T)} batches')

   
    self.current_position = 0

  
  def next_batch(self):
    B, T = self.B, self.T
    buf = self.tokens[self.current_position: self.current_position + B * T + 1]
    x = buf[:-1].view(B, T) 
    y = buf[1:].view(B, T) 

    
    self.current_position += B * T

    if self.current_position + ( B * T + 1) > len(self.tokens):
      self.current_position = 0
    
    return x, y

# --------------------------------------------------------------------------------------------------------
import time

torch.manual_seed(1337)
device = 'cpu'

if torch.cuda.is_available():
   torch.cuda.manual_seed(1337)
   device = 'cuda'

elif hasattr(torch.backends, 'mps') and torch.mps.is_available():
  torch.mps.manual_seed(1337)
  device = 'mps'

print(f'using device: {device}')

train_loader = DataLoaderLite(B=16, T=1024)
torch.set_float32_matmul_precision('high')

model = GPT(GPTConfig(vocab_size=50304))
model.to(device)
model = torch.compile(model)

optimizer = torch.optim.AdamW(model.parameters(), lr=3e-4)
for i in range(50):
  t0 = time.time()
  x, y = train_loader.next_batch()
  x, y = x.to(device), y.to(device)
  optimizer.zero_grad()

  with torch.autocast(device_type=device, dtype=torch.bfloat16):
   logits, loss = model(x, y)
  
  loss.backward()
  optimizer.step()
  # torch.mps.synchronize() # Waits for all kernels in all streams on a MPS device to complete.
  t1 = time.time()
  dt = (t1 - t0) * 1000
  tokens_per_sec = (train_loader.B * train_loader.T) / (t1 - t0)
  print(f'step: {i}, loss: {loss.item()}, dt: {dt:.2f}ms, tok/sec: {tokens_per_sec:.2f}')

import sys
sys.exit(0)
```

with torch.compile

Image

we can see that loss spits out nan

without torch.compile

Image

We can see that loss converges

### Versions

collect_env.py                                               100%[==============================================================================================================================================>]  30.38K  --.-KB/s    in 0.001s  

2026-01-06 08:12:53 (39.3 MB/s) - ‘collect_env.py’ saved [31107/31107]

zsh: command not found: #
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 26.2 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.6.3.2)
CMake version: Could not collect
Libc version: N/A

Python version: 3.13.5 (v3.13.5:6cb20a219a8, Jun 11 2025, 12:23:45) [Clang 16.0.0 (clang-1600.0.26.6)] (64-bit runtime)
Python platform: macOS-26.2-arm64-arm-64bit-Mach-O
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: N/A
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A
Caching allocator config: N/A

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.2
[conda] Could not collect

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by Isalia20**
Fixes #171764

Took me a while to figure out wth was going wrong.

Mini reproducer:
```python
import torch

# (uint / 65536) % non_power of 2, gives wrong result
lib = torch.mps.compile_shader('''
    kernel void func(device int* out, uint idx [[thread_position_in_grid]]) {
        out[idx] = (idx / 65536) % 6;
    }
''')

out = torch.empty(128, device='mps', dtype=torch.int32)
lib.func(out)

# Every value should be 0 since xindex/65536 == 0 for xindex in [0,127]
for i in [0, 5, 6, 7, 63, 64]:
    print(f"{i=} got {out[i].item()}")
```

Same purely in swift
```swift
import Metal

let device = MTLCreateSystemDefaultDevice()!
let queue = device.makeCommandQueue()!

let shaderSource = """
  kernel void func(device int* out [[buffer(0)]], uint idx [[thread_position_in_grid]]) {
    out[idx] = (idx / 65536) % 6;
  }
"""

let library = try device.makeLibrary(source: shaderSource, options: nil)
let function = library.makeFunction(name: "func")!
let pipeline = try device.makeComputePipelineState(function: function)

let count = 128
let buffer = device.makeBuffer(length: count * MemoryLayout<Int32>.stride, options: .storageModeShared)!

let cmdBuf = queue.makeCommandBuffer()!
let encoder = cmdBuf.makeComputeCommandEncoder()!
encoder.setComputePipelineState(pipeline)
encoder.setBuffer(buffer, offset: 0, index: 0)
encoder.dispatchThreads(
    MTLSizeMake(count, 1, 1),
    threadsPerThreadgroup: MTLSizeMake(min(count, pipeline.maxTotalThreadsPerThreadgroup), 1, 1)
)
encoder.endEncoding()
cmdBuf.commit()
cmdBuf.waitUntilCompleted()

let ptr = buffer.contents().bindMemory(to: Int32.self, capacity: count)
for i in [0, 5, 6, 7, 63, 64] {
    print("i=\(i) got \(ptr[i])")
}
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/codegen/mps.py [hunk 1]
```diff
@@ -81,13 +81,18 @@ def _print_FloorDiv(self, expr: sympy.Expr) -> str:
     def _print_ModularIndexing(self, expr: sympy.Expr) -> str:
         x, div, mod = expr.args
         x = self.doprint(x)
+        orig_div = div
+        orig_mod = mod
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Nit
```suggestion
        orig_div, orig_mod = div, mod
```

File: torch/_inductor/codegen/mps.py [hunk 2]
```diff
@@ -81,13 +81,18 @@ def _print_FloorDiv(self, expr: sympy.Expr) -> str:
...
         if div != 1:
             div = self.doprint(div)
             if expr.is_integer:
                 x = f"({x}) / ({div})"
             else:
                 x = f"metal::floor({x}) / ({div})"
         mod = self.doprint(mod)
+        # Workaround for Metal compiler bug with fused (x / A) % B, see PR 175481
+        if orig_div == 65536 and (orig_mod & (orig_mod - 1)) != 0:
+            return f"c10::metal::safe_mod({x}, {mod})"
```
Comment (malfet) [inline torch/_inductor/codegen/mps.py]: Why do you need this `orig_div` and `orig_mod` here? Why not make this check at the very beginning? Or may be a better change would be to rename `div` on line 87 to `div_`

**Review by malfet — APPROVED** [review]

File: c10/metal/utils.h [hunk 3]
```diff
@@ -189,6 +189,19 @@ inline common_dtype<T, U> floor_divide(T x, U y) {
...
+template <
+    typename T,
+    typename U,
+    ::metal::enable_if_t<
+        is_scalar_integral_v<T> && is_scalar_integral_v<U>,
+        bool> = true>
+inline common_dtype<T, U> safe_mod(T x, U y) {
+  volatile auto tmp = x;
+  return tmp % y;
+}
```
Comment (malfet) [inline c10/metal/utils.h]: Nit (no need to alias)
```suggestion
inline common_dtype<T, U> safe_mod(volatile T x, U y) {
  return x % y;
}
```

**Review by malfet — COMMENTED** [review]
Can you also please add a regression test for this particular problem to `test_mps` (`TestMetalLibrary` class), where one call `safe_mod` to gets all 0 for expected inputs
